tensordict/nn/storage.py

```diff
@@ -0,0 +1,684 @@
+        self.batch_added = False
+
+    @classmethod
+    def from_tensordict_pair(
+        cls,
+        source,
+        dest,
+        in_keys: List[NestedKey],
+        out_keys: List[NestedKey] | None = None,
+        storage_type: type = LazyDynamicStorage,
+        hash_module: Callable | None = None,
+    ):
+        """Creates a new TensorDictStorage from a pair of tensordicts (source and dest) using pre-defined rules of thumb.
+
+        Args:
+            source (TensorDict): An example of source tensordict, used as index in the storage.
+            dest (TensorDict): An example of dest tensordict, used as data in the storage.
+            in_keys (List[NestedKey]): a list of keys to use in the map.
+            out_keys (List[NestedKey]): a list of keys to return in the output tensordict.
+                All keys absent from out_keys, even if present in ``dest``, will not be stored
+                in the storage. Defaults to ``None`` (all keys are registered).
+            storage_type (type, optional): a type of tensor storage.
+                Defaults to :class:`~tensordict.nn.storage.LazyDynamicStorage`.
+                Other options include :class:`~tensordict.nn.storage.FixedStorage`.
+            hash_module (Callable, optional): a hash function to use in the :class:`~tensordict.nn.storage.QueryModule`.
+                Defaults to :class:`SipHash` for low-dimensional inputs, and :class:`~tensordict.nn.storage.RandomProjectionHash`
+                for larger inputs.
+
+        Examples:
+            >>> # The following example requires torchrl and gymnasium to be installed
+            >>> from tensordict.nn.storage import TensorDictStorage, RandomProjectionHash
+            >>> from torchrl.envs import GymEnv
+            >>> env = GymEnv("CartPole-v1")
+            >>> rollout = env.rollout(100)
+            >>> source, dest = r.exclude("next"), r.get("next")
+            >>> storage = TensorDictStorage.from_tensordict_pair(
+            ...     source, dest,
+            ...     in_keys=["observation", "action"],
+            ... )
+            >>> # maps the (obs, action) tuple to a corresponding next state
+            >>> storage[source] = dest
+            >>> storage[source]
+            TensorDict(
+                fields={
+                    done: Tensor(shape=torch.Size([35, 1]), device=cpu, dtype=torch.bool, is_shared=False),
+                    observation: Tensor(shape=torch.Size([35, 4]), device=cpu, dtype=torch.float32, is_shared=False),
+                    reward: Tensor(shape=torch.Size([35, 1]), device=cpu, dtype=torch.float32, is_shared=False),
+                    terminated: Tensor(shape=torch.Size([35, 1]), device=cpu, dtype=torch.bool, is_shared=False),
+                    truncated: Tensor(shape=torch.Size([35, 1]), device=cpu, dtype=torch.bool, is_shared=False)},
+                batch_size=torch.Size([35]),
+                device=None,
+                is_shared=False)
```

Comment:
I'd like to have a better, standalone example, but this one is nice bc it's a bit "real-life" and makes the point clear about the storage

cc @dtsaras for info

Reply:
> TensorDictStorage allows you to use a tensordict as an index, to build a bijective function
The bijective function is an interesting view of TensorDictStorage, though originally the reason I needed it in MCTS was to have a way to manipulate the output of a function (not necessarily the bijective function).

TensorDictModule is also a function, but it is difficult to change its output. I needed something like a memory/storage to save, retrieve, and *update* a value.

> For this feature to work we need to parse the tensordict used as index into a "real" long tensor index.
This is true for this implementation, but I hope we can extend this in the future so that we can support KNN search with the same abstraction, such as retrieving the K-closest items to the input. This can be quite useful for RAG LLM.

> TensorDict2Index and Index2TensorDict
I like these names, they are more intuitive 👍 

> The most crucial thing I can think of right now is what are the limitations and assumptions, and in which situations will this fail to be efficient.
To me, the biggest limitation is that it does not allow the gradient to flow to memory, at least in this implementation. The second issue could be performance-related, in particular, DynamicStorage and SipHash have both been implemented in CPU memory, so tensors have to move from GPU. (SipHash could have been implemented with PyTorch operations, so this can be alleviated in the future.) 

Finally, concurrency is another issue, especially in distributed training jobs.

> Are we limited in size, ie can we use billions of tensordicts as indices or will this break?
This a good question, it is quite interesting to see if we can wrap [FAISS](https://github.com/facebookresearch/faiss) in this api. 

> After giving a thorough look at all this I think that TensorDictMap is more appropriate

While I originally thought about this API as a memory/storage API (the main added value is the ease in *updating* a storage/memory value), I have no strong opinion here. Either name works.